lib/internal/util/inspect.js

```diff
@@ -1533,14 +1533,20 @@ function formatWithOptions(inspectOptions, ...args) {
           switch (nextChar) {
             case 115: // 's'
               const tempArg = args[++a];
-              if (typeof tempArg !== 'string' &&
-                  typeof tempArg !== 'function') {
+              if (typeof tempArg === 'object' && tempArg !== null &&
+                  (!hasOwnProperty(tempArg, 'toString') ||
```

Comment:
I think this line should be changed, because you generally also want to catch cases like

```js
class Foo {
  constructor(x,y) { this.x = x; this.y = y; }
  toString() { return `${this.x};${this.y}`; }
}

util.format('%s', new Foo())
```

… right? i.e. something like 

```suggestion
                  (tempArg.toString !== Object.prototype.toString ||
```

Reply:
That's a good point. The suggestion won't work for code from different realms though. What do you think about:

```js
String(tempArg.toString) !== 'function toString() { [native code] }'
```

Reply:
That would not catch native objects created by addons – that’s an edge case, but I think it’s a possibility. If you want to also account for different realms, I think `tempArg.toString !== tempArg.toString.__proto__.__proto__.toString` would work somewhat consistently (not ideal, of course, and needs checking for whether all these properties exist, but it’s the best thing that comes to my mind).

Reply:
This sadly does not work for all cases either, since the built-in `toString` function is not always identical to `Object.prototype.toString`. We can not even use `Object.prototype.toString`, since we use Primordials.

I struggle finding a solution that works for all cases.

Reply:
Can you give a realistic-ish example for which it doesn’t work? Like, sure, you can invalidate this check by e.g. changing the prototype of `Function.prototype`, but that seems like nothing that *we* should need to account for in all cases…

Reply:
```console
> a = [1,2,3]
[ 1, 2, 3 ]
> a.toString.__proto__.__proto__.toString === a.toString
false
```

Reply:
I mean … that’s happening because the `Array` class has a custom `toString()` method, similar to my example above – I guess we can’t get super-picky and decide which `toString()` methods are “good” and which are “bad”? I wouldn’t be surprised if `%s` with an Array would stringify it according to the standard rules…

Reply:
I changed the check to the following: either `toString` is an own property on the object __or__ the prototype of the constructor of the object has an own `toString` property and the constructor is not a built-in object (checked only with the constructor name to work in any realm).

That way we use the `toString` function in case it's defined by users and use `util.inspect` for all other cases. There are still edge cases (the native `toString` function could be replaced) but we can't work around all edge cases.